lib/pbench/server/database/models/users.py

```diff
@@ -5,10 +5,23 @@
 from flask_bcrypt import generate_password_hash
 from sqlalchemy import Column, DateTime, Enum, Integer, LargeBinary, String
 from sqlalchemy.orm import relationship, validates
+from sqlalchemy.orm.exc import NoResultFound
 
 from pbench.server.database.database import Database
 
 
+class AdminUserDeletion(Exception):
+    """
+    Attempt to delete an admin user.
```

Comment:
Shouldn't this be that the "last" admin user cannot be deleted?  Or are we saying that you have to first mark a user as not an admin before deleting?

Reply:
I've convinced myself that, at least until we expand the roles, a non-admin user can only delete itself, while an admin user can delete any user **_except_** itself.  If you want to get rid of an admin user, you need to do it from another admin user.  In this way, you can never delete the last admin.

We probably need to invoke similar logic in removing the admin role from a user....

Reply:
Yeah at the moment we are not dealing with checking whether the user is a last admin user (that will come when we have roles), basically if the user is an admin it can not be deleted, we have to mark a user as not an admin before deleting.

Reply:
I think it's a pain to have to figure out whether the target is the _last_ Admin user.  I think instead imposing the requirement that an Admin user not delete itself will save us from that pain, and it's simple to check.  I don't think we need any additional roles support before we consider or resolve this question.

In a similar vein, I don't think we should allow an Admin user to remove him/herself from the Admin role -- that's really equivalent to deleting the Admin user, in terms of trying to make sure that there is at least one left.

However, I also think we need to keep the mechanism separated from the policy:  at the database `User` level, I don't think there should be _any_ roles check; instead, it should be up to the caller to decide whether or not it is OK to delete the target user.

And, I think the restrictions on deleting users of various kinds should only be imposed on requests coming from the API -- the server-side CLI isn't subject to authorization checks, and I don't think it needs to be subject to deletion checks.  The API has the problem that it cannot be used to create an Admin user unless the request comes from an Admin user, so it behooves us to ensure that there is at least one Admin user left for the API to use when the deletion operation is initiated by the API.  However, the CLI has no such restrictions:  it will be used to create the initial Admin user, and, if it is used to delete the one and only remaining Admin user that is not a problem, because it can then be used to create a new one.

So, in the context of this PR, I believe that this discussion is beside the point (other than to highlight that perhaps this code change should be removed).